## Упражнения по библиотеке Numpy

In [2]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [13]:
def random_array():
    return np.random.randint(-5, 15, 15)



In [16]:
array = random_array()
print(*array)

array[(array > 3) & (array < 8)] *= -1
print(*array)



13 9 13 -3 2 -1 -3 5 14 2 7 0 -5 -4 9
13 9 13 -3 2 -1 -3 -5 14 2 -7 0 -5 -4 9


**2.** Заменить максимальный элемент случайного массива на 0

In [22]:
array2 = random_array()
print(*array2)
array2[array2.argmax()] = 0
print(*array2)

0 5 -1 9 3 3 8 7 13 -1 -2 0 -5 5 11
0 5 -1 9 3 3 8 7 0 -1 -2 0 -5 5 11


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [27]:
def decart(arr1, arr2):
    return np.array([(i,j) for i in arr1 for j in arr2])


print(*decart([1, 2, 3],[0, 1, 4]))

[1 0] [1 1] [1 4] [2 0] [2 1] [2 4] [3 0] [3 1] [3 4]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [45]:
A = np.random.randint(-5, 15, (8, 3))
B = np.random.randint(-5, 15, (2, 2))

print(A, end='\n\n\n')
print(B, end='\n\n\n')

compared_matrix = (A[..., np.newaxis, np.newaxis] == B)
result = A[np.all(np.any(compared_matrix, axis =(1, 3)), axis = 1)]
print(result)



[[ 8  1 -3]
 [ 6 -3 14]
 [14  6 -2]
 [ 5 11 -3]
 [ 7 12 -5]
 [ 4 -1 -2]
 [ 0  1  6]
 [13  0  1]]


[[ 9 -2]
 [ 0  6]]


[[14  6 -2]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [75]:
A = np.random.randint(1, 3, (10, 3))
print(A, end='\n\n\n\n\n')

equal = np.logical_and.reduce(A[:,1:] == A[:,:-1], axis=1)
print(A[~equal])

[[2 2 1]
 [1 1 2]
 [2 1 1]
 [2 1 1]
 [2 2 2]
 [2 2 1]
 [2 2 1]
 [2 1 1]
 [1 2 2]
 [1 2 2]]




[[2 2 1]
 [1 1 2]
 [2 1 1]
 [2 1 1]
 [2 2 1]
 [2 2 1]
 [2 1 1]
 [1 2 2]
 [1 2 2]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [93]:
A = np.random.randint(1, 3, (10, 3))
print(A, end='\n\n\n\n\n')

B = np.unique(A, axis=0)
print(B)

[[2 2 2]
 [1 2 2]
 [2 2 2]
 [2 2 2]
 [1 2 1]
 [1 1 2]
 [1 1 1]
 [1 1 1]
 [2 2 2]
 [2 1 2]]




[[1 1 1]
 [1 1 2]
 [1 2 1]
 [1 2 2]
 [2 1 2]
 [2 2 2]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [104]:
array = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
mult = 1

for i in range(len(array)):
    for j in range(len(array[1])):
        if i == j and array[i][j] != 0:
            mult *= array[i][j]
            
print (mult)            
            

3


In [110]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) 
mult = np.prod(np.diag(X)[np.diag(X) != 0])

print(mult)

3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [116]:
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]

print(set(x) == set(y))

x_ = np.array([1, 2, 2, 4])
y_ = np.array([4, 2, 1, 2])
print(np.all((np.sort(x)==np.sort(y))))

True
True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [136]:
array = [6, 2, 0, 3, 0, 0, 5, 7, 0]
flag = False
for i in range(len(array)-1):
    if array[i] == 0:
        if flag:
            maximum = max(array[i+1], maximum)
        else:
            maximum = array[i+1]
            flag = True
print (maximum) 


x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) 
print(np.max(x[1:][(x==0)[:-1]]))

    

5
5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [151]:
array = [2, 2, 2, 3, 3, 3, 5]
return_vector1 = list(set(array))
counter = [0 for i in range(len(return_vector1))]

for i in range(len(return_vector1)):
    for j in range(len(array)):
        if return_vector1[i] == array[j]:
            counter[i] += 1 

print(return_vector1)
print(counter)

array =  np.array([2, 2, 2, 3, 3, 3, 5])
return_vector2 = np.unique(array)
counter = np.bincount(array)[np.bincount(array) != 0]

print(return_vector2)
print(counter)

[2, 3, 5]
[3, 3, 1]
[2 3 5]
[3 3 1]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [254]:
import time

#без numpy
X = [[1, 5], [1, 10], [2, 7]]
Y = [[2, 7], [1, 4], [3, 1]]
dist_x = []
dist_y = []

start_time_1 = time.time()
for i in range(len(X)):
    for j in range(len(X[1])-1):
        dist_x.append(abs(X[i][j] - X[i][j+1]))
 
for i in range(len(Y)):
    for j in range(len(Y[1])-1):
        dist_y.append(abs(Y[i][j] - Y[i][j+1]))

        
dist_nnumpy = []
for i in range(len(dist_x)):
    dist_nnumpy.append(pow(pow(dist_x[i],2) + pow(dist_y[i], 2), 0.5))

print('Plain Python realization')
print(dist_nnumpy)
print('Took', (time.time() - start_time_1), 'ms', end='\n\n')

#numpy
start_time_2 = time.time()
dist_numpy = (np.array([dist_x, dist_y])**2).sum(axis=0)**0.5

print('Numpy realization')
print(dist_numpy)
print('Took', (time.time() - start_time_2), 'ms', end='\n\n')


Plain Python realization
[6.4031242374328485, 9.486832980505138, 5.385164807134504]
Took 0.0008437633514404297 ms

Numpy realization
[6.40312424 9.48683298 5.38516481]
Took 0.0005502700805664062 ms



_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [152]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [153]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [155]:
average_calories = calorie_stats.mean()
print(average_calories)

106.88311688311688

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [158]:
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [160]:
median_calories = np.median(calorie_stats)
print(median_calories)

110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [180]:
print(np.percentile(calorie_stats, 100))
print(np.percentile(calorie_stats, 50))
print(np.percentile(calorie_stats, 25))
print(np.percentile(calorie_stats, 12.5))
print(np.percentile(calorie_stats, 6.25))
print(np.percentile(calorie_stats, 3.125))
print(np.percentile(calorie_stats, 3.2895))

nth_percentile = 3.2895

160.0
110.0
100.0
90.0
77.5
57.5
60.0004


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [193]:
more_calories = len(calorie_stats[calorie_stats > 60]) * 100 / len(calorie_stats)
print(more_calories)

96.1038961038961


77

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [191]:
calorie_std = np.std(calorie_stats)
print(calorie_std.round(2))

19.36


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

# Что мы имеем:
1. Наши хлопья менее калорийные, чем 96,1% других хлопьев
2. Среднее значение по калорийности 106.8 ккал, а стандартное отклонение 19.36, то есть
   наши хлопья содержат калорий меньше, чем среднее значение - стандартное отклонение
   
# Как использовать?
1. Можно сегментировать нашу ЦА и указатб, что продукт подходит для людей, которые поддерживают свой вес.
Вероятнее всего родители захотят кормить детей более питательными хлопьями.
2. Так как большая часть людей не знают статистики, то следует опираться на самые простые и понятные цифры. Можно использовать 96,1% как главный слоган рекламы, это покажет, чем наши хлопья отличаются от других 77 таких же продуктов.